File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
import sys
sys.setrecursionlimit(10000)


def dfs(x, y, count):
    graph[x][y] = 1
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        if (0 <= nx < m) and (0 <= ny < n) and graph[nx][ny] == 0:
            count = dfs(nx, ny, count + 1)
    return count


dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

m, n, k = map(int, input().split(" "))
graph = [[0] * n for _ in range(m)]
result = []
for _ in range(k):
    x1, y1, x2, y2 = map(int, input().split(" "))
    for i in range(y1, y2):
        for j in range(x1, x2):
            graph[i][j] = 1
for i in range(m):
    for j in range(n):
        if graph[i][j] == 0:
            result.append(dfs(i, j, 1))
print(len(result))
print(*sorted(result))

File: 백준/Silver/2583. 영역 구하기/영역 구하기.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val (m, n, k) = reader.readLine().split(" ").map { it.toInt() }
    val visited = Array(m) { BooleanArray(n) { false } }
    repeat(k) {
        val (x1, y1, x2, y2) = reader.readLine().split(" ").map { it.toInt() }
        for (i in y1 until y2) {
            for (j in x1 until x2) {
                visited[i][j] = true
            }
        }
    }

    val result = mutableListOf<Int>()
    for (i in 0 until m) {
        for (j in 0 until n) {
            if (!visited[i][j]){
                result.add(dfs(i, j, visited))
            }
        }
    }
    result.sort()

    println(result.size)
    println(result.joinToString(" "))
}

fun dfs(x: Int, y: Int, visited: Array<BooleanArray>): Int {
    var count = 1
    visited[x][y] = true
    
    val dx = arrayOf(0, 0, 1, -1)
    val dy = arrayOf(1, -1, 0, 0)
    
    for (i in 0 until 4) {
        val nx = x + dx[i]
        val ny = y + dy[i]
        if (nx in visited.indices && ny in visited[0].indices && !visited[nx][ny]) {
            count += dfs(nx, ny, visited)
        }
    }
    return count
}